File: backend/funcionario/src/main/java/com/ms/funcionario/controllers/FuncionarioController.java
package com.ms.funcionario.controllers;

import com.ms.funcionario.dto.FuncionarioDTO;
import com.ms.funcionario.services.FuncionarioService;
import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/funcionarios")
public class FuncionarioController {

    @Autowired
    private FuncionarioService service;

    @GetMapping
    public List<FuncionarioDTO> listAll() {
        return service.listarFuncionarios();
    }

    @GetMapping("/{id}")
    public ResponseEntity<FuncionarioDTO> getById(@PathVariable Long id) {
        try {
            FuncionarioDTO dto = service.findById(id);
            
            return ResponseEntity.ok(dto);
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    @PostMapping
    public ResponseEntity<FuncionarioDTO> createCliente(@RequestBody FuncionarioDTO funcionarioDTO) {
        FuncionarioDTO saved = service.saveFuncionario(funcionarioDTO);
        
        return ResponseEntity.status(HttpStatus.CREATED).body(saved);
    }

    @PutMapping("/{id}")
    public ResponseEntity<FuncionarioDTO> updateFuncionario(@PathVariable Long id, @RequestBody FuncionarioDTO funcionarioDTO) {
        try {
            FuncionarioDTO updated = service.updateFuncionario(id, funcionarioDTO);

            return ResponseEntity.ok(updated);
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(null);
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteFuncionario(@PathVariable Long id) {
        try {
            service.deleteFuncionario(id);

            return ResponseEntity.noContent().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }
}

File: backend/funcionario-service/src/main/kotlin/br/ufpr/dac/funcionario_service/resource/FuncionarioController.kt
package br.ufpr.dac.funcionario_service.resource

import utils.dto.FuncionarioOutputDTO
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.bind.annotation.DeleteMapping
import utils.dto.FuncionarioInputDTO
import org.springframework.beans.factory.annotation.Value

@RestController
@RequestMapping("/v1/funcionarios")
@CrossOrigin(origins = ["\${gateway.url}"])
class FuncionarioController(private val service: FuncionarioService) {

    @GetMapping
    fun getFuncionarios(): ResponseEntity<List<FuncionarioOutputDTO>> {
        val funcionarios = service.getAllFuncionarios()
        return ResponseEntity.ok().body(funcionarios)
    }

    @GetMapping("/{id}")
    fun getFuncionarioById(@PathVariable id: Long): ResponseEntity<FuncionarioOutputDTO> {
        val funcionario = service.getFuncionarioById(id)
        return ResponseEntity.ok().body(funcionario)
    }

    @PutMapping("/{id}")
    fun updateFuncionario(
        @Valid @RequestBody funcionario: FuncionarioInputDTO, @PathVariable id: Long
    ): ResponseEntity<FuncionarioOutputDTO> {
        val updatedFuncionario = service.updateFuncionario(id, funcionario)
        return ResponseEntity.ok().body(updatedFuncionario)
    }

    @PostMapping
    fun createFuncionario(@Valid @RequestBody funcionario: FuncionarioInputDTO): ResponseEntity<FuncionarioOutputDTO> {
        return try {
            val savedFuncionario = service.saveFuncionario(funcionario)
            ResponseEntity.status(HttpStatus.CREATED).body(savedFuncionario)
        } catch (e: IllegalArgumentException) {
            // Caso de erro de CPF inválido
            throw IllegalArgumentException(e.message)
        }
    }

    @DeleteMapping("/{id}")
    fun deleteFuncionario(@PathVariable id: Long): ResponseEntity<FuncionarioOutputDTO> {
        val deactivatedFuncionario = service.deactivateFuncionario(id)
        return ResponseEntity.ok(deactivatedFuncionario)
    }
}


